core/meterer/meterer_test.go

```diff
@@ -153,3 +174,321 @@ func teardown() {
+	assert.NoError(t, err)
+	err = mt.MeterRequest(ctx, *header)
+	assert.ErrorContains(t, err, "quorum number mismatch")
+
+	// test non-existent account
+	unregisteredUser, err := crypto.GenerateKey()
+	if err != nil {
+		t.Fatalf("Failed to generate key: %v", err)
+	}
+	header, err = auth.ConstructPaymentMetadata(&signer, 1, 0, 1000, []uint8{0, 1, 2}, unregisteredUser)
+	assert.NoError(t, err)
+	err = mt.MeterRequest(ctx, *header)
+	assert.ErrorContains(t, err, "failed to get active reservation by account: reservation not found")
+
+	// test invalid bin index
+	header, err = auth.ConstructPaymentMetadata(&signer, binIndex, 0, 2000, quoromNumbers, privateKey1)
+	assert.NoError(t, err)
+	err = mt.MeterRequest(ctx, *header)
+	assert.ErrorContains(t, err, "invalid bin index for reservation")
+
+	// test bin usage metering
+	accountID := crypto.PubkeyToAddress(privateKey2.PublicKey).Hex()
+	for i := 0; i < 9; i++ {
+		dataLength := 20
+		header, err = auth.ConstructPaymentMetadata(&signer, binIndex, 0, uint32(dataLength), quoromNumbers, privateKey2)
+		assert.NoError(t, err)
+		err = mt.MeterRequest(ctx, *header)
+		assert.NoError(t, err)
+		item, err := dynamoClient.GetItem(ctx, "reservations", commondynamodb.Key{
+			"AccountID": &types.AttributeValueMemberS{Value: accountID},
+			"BinIndex":  &types.AttributeValueMemberN{Value: strconv.Itoa(int(binIndex))},
+		})
+		assert.NoError(t, err)
+		assert.Equal(t, accountID, item["AccountID"].(*types.AttributeValueMemberS).Value)
+		assert.Equal(t, strconv.Itoa(int(binIndex)), item["BinIndex"].(*types.AttributeValueMemberN).Value)
+		assert.Equal(t, strconv.Itoa(int((i+1)*dataLength)), item["BinUsage"].(*types.AttributeValueMemberN).Value)
+
+	}
+	// frist over flow is allowed
+	header, err = auth.ConstructPaymentMetadata(&signer, binIndex, 0, 25, quoromNumbers, privateKey2)
+	assert.NoError(t, err)
+	err = mt.MeterRequest(ctx, *header)
+	assert.NoError(t, err)
+	overflowedBinIndex := binIndex + 2
+	item, err := dynamoClient.GetItem(ctx, "reservations", commondynamodb.Key{
+		"AccountID": &types.AttributeValueMemberS{Value: accountID},
+		"BinIndex":  &types.AttributeValueMemberN{Value: strconv.Itoa(int(overflowedBinIndex))},
+	})
+	assert.NoError(t, err)
+	assert.Equal(t, accountID, item["AccountID"].(*types.AttributeValueMemberS).Value)
+	assert.Equal(t, strconv.Itoa(int(overflowedBinIndex)), item["BinIndex"].(*types.AttributeValueMemberN).Value)
+	assert.Equal(t, strconv.Itoa(int(5)), item["BinUsage"].(*types.AttributeValueMemberN).Value)
+
+	// second over flow
+	header, err = auth.ConstructPaymentMetadata(&signer, binIndex, 0, 1, quoromNumbers, privateKey2)
+	assert.NoError(t, err)
+	err = mt.MeterRequest(ctx, *header)
+	assert.ErrorContains(t, err, "bin has already been filled")
+
+	// overwhelming bin overflow for empty bins (assuming all previous requests happened within 1 reservation window)
+	header, err = auth.ConstructPaymentMetadata(&signer, binIndex-1, 0, 1000, quoromNumbers, privateKey2)
```

Comment:
Shoud we also have a passing test for `binIndex-1`?

Reply:
yep! added it and ready for review